server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java

```diff
@@ -65,35 +71,78 @@
+    ) {
+        this(from, size, trackTotalHitsUpTo, searchTimeProvider, aggReduceContextBuilder, clusters, progressListener, -1);
+    }
 
+    /**
+     *
+     * @param from
+     * @param size
+     * @param trackTotalHitsUpTo
+     * @param searchTimeProvider
+     * @param aggReduceContextBuilder
+     * @param clusters
+     * @param progressListener
+     * @param numTotalClusters Specify the total number of clusters to turn on an optimization where an incremental
+     *                         merge will not be done in the add method for the last search response to come in
+     *                         (since the user of the Merger object is going to call {getMergedResponse} after
+     *                         all have been added.
+     */
     SearchResponseMerger(
         int from,
         int size,
         int trackTotalHitsUpTo,
         SearchTimeProvider searchTimeProvider,
-        AggregationReduceContext.Builder aggReduceContextBuilder
+        AggregationReduceContext.Builder aggReduceContextBuilder,
+        SearchResponse.Clusters clusters,
+        SearchProgressListener progressListener,
+        int numTotalClusters
     ) {
         this.from = from;
         this.size = size;
         this.trackTotalHitsUpTo = trackTotalHitsUpTo;
         this.searchTimeProvider = Objects.requireNonNull(searchTimeProvider);
         this.aggReduceContextBuilder = aggReduceContextBuilder; // might be null if there are no aggregations
+        this.clusters = clusters;
+        this.progressListener = progressListener;
+        this.numExpectedResponses = numTotalClusters;
     }
 
     /**
      * Add a search response to the list of responses to be merged together into one.
-     * Merges currently happen at once when all responses are available and {@link #getMergedResponse(Clusters)} )} is called.
+     * Merges currently happen at once when all responses are available and {@link #getMergedResponse()} )} is called.
      * That may change in the future as it's possible to introduce incremental merges as responses come in if necessary.
      */
-    void add(SearchResponse searchResponse) {
+    synchronized void add(SearchResponse searchResponse) {
         assert searchResponse.getScrollId() == null : "merging scroll results is not supported";
         searchResponses.add(searchResponse);
+        // sync-search uses a NOOP progress listener so only do incremental reductions for async-searches
+        // optimization: if all the responses have now been received (numExpectedResponses == numResponses()), then do not
+        // do an incremental merge, as the CCSActionListener will call getMergedResponse to create the final response
+        // after this method returns
+        if (progressListener != null && progressListener != SearchProgressListener.NOOP && numExpectedResponses != numResponses()) {
+            progressListener.notifyCcsReduce(getMergedResponse());
+        }
```

Comment:
It feels a bit weird to carry the progress listener in SearchResponseMerger. Would it be an option to call nofityCcsReduce after we call SearchResponseMerger#add ?

Reply:
It's  not a huge deal on my end, but I think that SearchResponseMerger is in principle a lightweight class that is able to merge search responses together, without knowing too much about the context in which it does that. Making it aware of progress listener makes it more complex, especially with the null/noop check. I find that calling notifyCCSReduce outside of it would make sure that we call that only when needed and removes the need for the check. I would not feel bad about CCSACtionListener becoming aware of SearchProgressListener instead. It feels actually more natural to have a listener notify another listener, compared to a sync component notifying a listener of some of its actions.

Reply:
To add to that, I did not realize that `add` calls` getMergedResponse`. I focused on the conditional and not on what it does within the if block :)

This looks like bad practice to me. I think that add should be called from a consumer, and the same as getMergedResponse.

Reply:
OK. I will close this PR and start a new one.